### PR TITLE
client: Accept structured parameters in non-object operations

### LINKED ISSUE #92: Use parameter structure in all client methods
As the part of api-go to sdk-go client transition, we wanted to unify all client methods. We did this with response values (all of them include status now). It is time to do this with invocation parameters.

## Proposals
All client methods should have similar signature: `MethodName(ctx, prm) (res, err)`. Parameter structure should provide setters for all required data, including bearer tokens, session tokens, keys, and other invocation modifiers.

### PR DESCRIPTION
**Pull request by cthulhu-rider**
* Closes #92

Define `XPrm` type for each `X` client operation which structures
parameters. Export setters of each parameterized value. Emphasize that
some parameters are required. Make the client panic when the parameters
are incorrectly set. Get rid of vadiadic call options and `CallOption`
type. Improve documentation of client behavior.

Signed-off-by: Leonard Lyubich <leonard@nspcc.ru>

## TODO

- [x] non-object ops
- [x] doc status errors

## TBD
Object ops (separate PR)

### CODE REVIEW

File: client/accounting.go [hunk 1]
```diff
@@ -7,76 +7,122 @@ import (
...
+// GetBalance requests current balance of the NeoFS account.
+//
+// Immediately panics if parameters are set incorrectly (see GetBalancePrm docs).
+// Context is required and must not be nil. It is used for network communication.
+//
+// Exactly one return value is non-nil. Server status return is returned in GetBalanceRes.
+// Reflects all internal errors in second return value (transport problems, response processing, etc.).
+func (c *clientImpl) GetBalance(ctx context.Context, prm GetBalancePrm) (*GetBalanceRes, error) {
+	// check context
+	if ctx == nil {
```
Comment (alexvanin) [inline client/accounting.go:70]: Won't it panic anyway if `ctx == nil`?

**Review by alexvanin — COMMENTED** [review]
By the way, maybe split this PR into two parts:
- everything except object service methods,
- object service methods?

Object service methods use input and output streams and much more complex in general. It will be much easier to review and discuss details if it will be split, I think.

File: client/accounting.go (hunk 1 above)
Comment (cthulhu-rider) [reply client/accounting.go:70]: It will, but much later and caused by NPE which is not clear enough.

File: client/accounting.go [hunk 2]
```diff
@@ -2,82 +2,102 @@ package client
...
-	// apply all available options
-	callOptions := c.defaultCallOptions()
-
-	for i := range opts {
-		opts[i](callOptions)
+//
+// Immediately panics if parameters are set incorrectly (see GetBalancePrm docs).
+// Context is required and must not be nil. It is used for network communication.
+//
+// Exactly one return value is non-nil. Server status return is returned in GetBalanceRes.
```
Comment (carpawell) [inline client/accounting.go:54]: it conflicts a little with the paragraph above (that has info about `WithNeoFSErrorParsing`)

File: client/container.go [hunk 3]
```diff
@@ -352,149 +421,199 @@ func (c *Client) DeleteContainer(ctx context.Context, id *cid.ID, opts ...CallOp
...
+//
+// Operation is asynchronous and no guaranteed even in the absence of errors.
+// The required time is also not predictable.
+//
+// Success can be verified by reading by identifier (see EACL).
+//
+// Immediately panics if parameters are set incorrectly (see SetEACLPrm docs).
+// Context is required and must not be nil. It is used for network communication.
+//
+// Exactly one return value is non-nil. Server status return is returned in ContainerDeleteRes.
```
Comment (carpawell) [inline client/container.go:595]:
```suggestion
// Exactly one return value is non-nil. Server status return is returned in SetEACLRes.
```

**Review by carpawell — COMMENTED** [review]
why `panic` was chosen as behavior on invalid parameters?

File: pool/pool.go [hunk 4]
```diff
@@ -898,14 +952,21 @@ func (p *pool) WaitForContainerPresence(ctx context.Context, cid *cid.ID, pollPa
 	defer ticker.Stop()
 	wdone := wctx.Done()
 	done := ctx.Done()
+
+	var cliPrm client.ContainerGetPrm
+
+	if cid != nil {
+		cliPrm.SetContainer(*cid)
+	}
```
Comment (alexvanin) [inline pool/pool.go:960]: As soon as client library throws panics, I suppose it is okay to not check `nil` there and just go straight `cliPrm.SetContainer(*cid)`. It is invalid case when `cid` is nill.

The same comment applied for others `nil` check on mandatory arguments in pool package.

File: client/opts.go [hunk 5]
```diff
@@ -51,8 +51,6 @@ type (
 	v2SessionReqInfo struct {
 		addr *refs.Address
 		verb v2session.ObjectSessionVerb
-
-		exp, nbf, iat uint64
```
Comment (alexvanin) [inline client/opts.go:55]: Why those are not required anymore? And the same as lifetime in session token.

File: client/container.go [hunk 6]
```diff
@@ -167,177 +169,244 @@ func (x *ContainerGetRes) setContainer(cnr *container.Container) {
 	x.cnr = cnr
 }
 
-// GetContainer receives container structure through NeoFS API call.
+// GetContainer reads NeoFS container from by ID.
```
Comment (alexvanin) [inline client/container.go:172]:
```suggestion
// GetContainer reads container from NeoFS by ID.
```

**Review by alexvanin — COMMENTED** [review]
LGTM, did some questions in the comments and will approve after.

Also, what do you think about returning parameter structure in setters? It seemed quite nice in object service parameters when you could do
```
prm.SetA(..).
  SetB(..).
  SetC(..)
```

**Comment by cthulhu-rider** [comment]
> why `panic` was chosen as behavior on invalid parameters?

Because all the parameters are strictly documented. SDK can deal with incorrect parameters as well as its users. If we'll decide to relax parameter requirements - this will be a compatible change.

File: pool/pool.go (hunk 4 above)
Comment (cthulhu-rider) [reply pool/pool.go:960]: W/o nil-check here program will panic with generic NPE, while `Client` panics with the exact reason. In future I hope we will accept `cid` by value, not a pointer.

File: client/opts.go (hunk 5 above)
Comment (cthulhu-rider) [reply client/opts.go:55]: These fields were never set. I decided to decrease client's business logic and leave token composition on the user side.

**Review by alexvanin — APPROVED** [review]
There is a opinion that libraries should never panic, while application can. But I agree, that we can change this behavior later in compatible way if there will be high demand on that.

Also, what about setters that return `Prm`? I think they are quite pretty and already used in object service parameters: `Prm.Set(A).Set(B).Set(C)`

**Comment by cthulhu-rider** [comment]
> Also, what about setters that return `Prm`? I think they are quite pretty and already used in object service parameters: `Prm.Set(A).Set(B).Set(C)`

If this approach is convenient, we can add `With...` methods as an alternative.
Like? @carpawell @fyrchik

**Comment by alexvanin** [comment]
> > Also, what about setters that return `Prm`? I think they are quite pretty and already used in object service parameters: `Prm.Set(A).Set(B).Set(C)`
> 
> If this approach is convenient, we can add `With...` methods as an alternative. Like? @carpawell @fyrchik

As for me, `With` prefix is more associated with optional parameters, and here we have required parameters as well. Anyway, this can be done later as well without compatibility break.

**Pull request merged**